cpp/src/stream_compaction/distinct.cu

```diff
@@ -37,18 +32,161 @@
+#include <thrust/uninitialized_fill.h>
 
 #include <utility>
 #include <vector>
 
 namespace cudf {
 namespace detail {
+
+namespace {
+/**
+ * @brief A functor to perform reduction on row indices.
+ *
+ * A reduction operation will be performed on each group of rows that are compared equal.
+ */
+template <typename MapDeviceView, typename KeyHasher, typename KeyComparator>
+struct reduce_index_fn {
+  MapDeviceView const d_map;
+  KeyHasher const d_hasher;
+  KeyComparator const d_eqcomp;
+  duplicate_keep_option const keep;
+  size_type* const d_output;
+
+  reduce_index_fn(MapDeviceView const& d_map,
+                  KeyHasher const& d_hasher,
+                  KeyComparator const& d_eqcomp,
+                  duplicate_keep_option const keep,
+                  size_type* const d_output)
+    : d_map{d_map}, d_hasher{d_hasher}, d_eqcomp{d_eqcomp}, keep{keep}, d_output{d_output}
+  {
+  }
+
+  __device__ void operator()(size_type const idx) const
+  {
+    if (keep == duplicate_keep_option::KEEP_FIRST) {
+      // Store the smallest index of all rows that are equal.
+      atomicMin(&get_output(idx), idx);
+    } else if (keep == duplicate_keep_option::KEEP_LAST) {
+      // Store the greatest index of all rows that are equal.
+      atomicMax(&get_output(idx), idx);
+    } else {
+      // Count the number of rows that are equal to the row having its index inserted.
+      atomicAdd(&get_output(idx), size_type{1});
+    }
+  }
+
+ private:
+  __device__ size_type& get_output(size_type const idx) const
+  {
+    // Here we don't check `iter` validity for performance reason, assuming that it is always
+    // valid because all input `idx` values have been fed into `map.insert()` before.
+    auto const iter = d_map.find(idx, d_hasher, d_eqcomp);
+
+    // Only one index value of the duplicate rows could be inserted into the map.
+    // As such, looking up for all indices of duplicate rows always returns the same value.
+    auto const inserted_idx = iter->second.load(cuda::std::memory_order_relaxed);
+
+    // All duplicate rows will have concurrent access to the same output slot.
+    return d_output[inserted_idx];
+  }
+};
+
+}  // namespace
+
+rmm::device_uvector<size_type> get_distinct_indices(table_view const& input,
+                                                    duplicate_keep_option keep,
+                                                    null_equality nulls_equal,
+                                                    rmm::cuda_stream_view stream,
+                                                    rmm::mr::device_memory_resource* mr)
+{
+  if (input.num_rows() == 0 or input.num_columns() == 0) {
+    return rmm::device_uvector<size_type>(0, stream, mr);
+  }
+
+  auto const preprocessed_input =
+    cudf::experimental::row::hash::preprocessed_table::create(input, stream);
+  auto const has_null = nullate::DYNAMIC{cudf::has_nested_nulls(input)};
+
+  auto map = hash_map_type{compute_hash_table_size(input.num_rows()),
```

Comment:
This `hash_map_type` alias makes it difficult to see what is going on here. Is it a `static_map`? `static_multimap`? What are the key/value types? etc. 

It'd be fine if it was defined in this function, or even in this header. But the fact that I have to go hunting for it in other headers is disruptive. 

Suffice to say, I think it hurts more than helps.

Reply:
Agree. So I'm going to define it here. But please be aware that such type alias is still being used in other places in cudf.